src/WebformHelper.php

```diff
@@ -0,0 +1,274 @@
+
+        foreach ($routes as $route) {
+          $parameters = [];
+
+          if ('rest.webform_rest_submit.POST' !== $route) {
+            $parameters['webform_id'] = $webform->id();
+          }
+          $uuidPlaceholder = '{uuid}';
+          if ($requireUuid($route)) {
+            $parameters['uuid'] = $uuidPlaceholder;
+          }
+          $parameters['api-key'] = $apiKey;
+
+          $url = Url::fromRoute($route, $parameters, ['absolute' => TRUE]);
+          $form['third_party_settings']['os2forms']['os2forms_rest_api']['api_info']['endpoints_test']['links'][$route] = [
+            '#type' => 'link',
+            '#title' => str_replace(urlencode($uuidPlaceholder), $uuidPlaceholder, $url->toString()),
+            '#url' => $url,
+            '#prefix' => '<li>',
+            '#suffix' => '</li>',
+          ];
+        }
+      }
+    }
+  }
+
+  /**
+   * Get webform by id or submission uuid.
+   *
+   * If submission uuid is specified (i.e. not null), the submission's webform's
+   * id must match the specified webform id.
+   *
+   * @return \Drupal\webform\WebformInterface|null
+   *   The webform if found.
+   */
+  public function getWebform(string $webformId, string $submissionUuid = NULL): ?WebformInterface {
+    if (NULL !== $submissionUuid) {
+      $storage = $this->entityTypeManager->getStorage('webform_submission');
+      $submissionIds = $storage
+        ->getQuery()
+        ->accessCheck(FALSE)
+        ->condition('uuid', $submissionUuid)
+        ->execute();
+      $submission = $storage->load(array_key_first($submissionIds));
+
+      if (NULL === $submission) {
+        return NULL;
+      }
+
+      assert($submission instanceof WebformSubmissionInterface);
+      $webform = $submission->getWebform();
+      if ($webformId !== $webform->id()) {
+        return NULL;
+      }
+
+      return $webform;
+    }
+
+    return $this->entityTypeManager
+      ->getStorage('webform')
+      ->load($webformId);
+  }
+
+  /**
+   * Get users allowed to access a webform's data.
+   *
+   * @return \Drupal\user\UserInterface[]|array
+   *   The users.
+   */
+  public function getAllowedUsers(WebformInterface $webform): array {
+    $settings = $webform->getThirdPartySetting('os2forms', 'os2forms_rest_api');
+    $allowedUserIds = $settings['allowed_users'] ?? [];
+
+    return $this->loadUsers($allowedUserIds);
+  }
+
+  /**
+   * Load users.
+   *
+   * @phpstan-param array<int, mixed> $spec
+   * @phpstan-return array<int, mixed>
+   */
+  private function loadUsers(array $spec): array {
+    return $this->entityTypeManager
+      ->getStorage('user')
+      ->loadMultiple(array_column($spec, 'target_id'));
+  }
+
+  /**
+   * Implements hook_file_download().
+   *
+   * @phpstan-return int|array<string, string>|null
+   */
+  public function fileDownload(string $uri) {
+    $request = \Drupal::request();
+
+    if ($user = $this->keyAuth->authenticate($request)) {
```

Comment:
does not that mean, that if we don't receive any user here - meaning `$user` is `NULL`, we skip any further code and return `NULL`. 
Returning NULL is the same as stopping intercepting file download

Reply:
I understand your concern here, but if I understand https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21File%21file.api.php/function/hook_file_download/9.x correctly, returning `NULL` basically states that we don't care about the file permissions and lets others decide. We can only deny access to API users authenticated via https://www.drupal.org/project/key_auth. Access for other users (including anonymous users) must be handled by other modules so I think it's correct to return `NULL` here.

Reply:
i don't quite agree with that. If a webform has defined a list of the users, that are able to perform some actions to it, that by definition means anonym user is not within allowed list.

To respect that anonym user should be forbidden access by this function, and not hope that some other function after that will do it.
In other words i mean that forbidding access to anonym user is this function responsibility.

Reply:
This function is only used to impose additional restrictions on which users can access (private) files attached to webform submissions. Therefore this function can only deny access and granting access is the responsibility of others. 

This is a bit weird and I cannot remember why we decided that, by default, all authenticated users should have access to all webforms and submissions with the possibility to limit the list of users with access. This may be a wrong decision and we may have to change this in the near future and explicitly grant access to REST resources.